runtime-attach/src/main/java/io/opentelemetry/contrib/attach/RuntimeAttach.java

```diff
@@ -7,22 +7,82 @@
 
 import java.io.File;
 import java.lang.management.ManagementFactory;
+import java.util.logging.Logger;
 import net.bytebuddy.agent.ByteBuddyAgent;
 
+/** This class allows you to attach the OTel agent for Java at runtime. */
 public final class RuntimeAttach {
 
+  private static final Logger LOGGER = Logger.getLogger(RuntimeAttach.class.getName());
+  private static final String AGENT_ENABLED_PROPERTY = "otel.javaagent.enabled";
+  private static final String AGENT_ENABLED_ENV_VAR = "OTEL_JAVAAGENT_ENABLED";
+  static final String MAIN_THREAD_CHECK_PROP = "otel.javaagent.runtimeattach.mainthreadcheck";
+
+  /**
+   * Attach the OTel agent for Java to the current JVM. The attachment must be requested at the
+   * beginning of the main method.
+   */
   public static void attachJavaagentToCurrentJVM() {
-    if (agentIsDisabled()) {
+    if (!shouldAttach()) {
       return;
     }
+
     File javaagentFile = AgentFileLocator.locateAgentFile();
     ByteBuddyAgent.attach(javaagentFile, getPid());
+
+    if (!agentIsAttached()) {
+      LOGGER.warning("Agent was not attached. An unexpected issue has happened.");
+    }
+  }
+
+  private static boolean shouldAttach() {
+    if (agentIsDisabledWithProp()) {
+      LOGGER.warning("Agent was disabled with " + AGENT_ENABLED_PROPERTY + " property.");
+      return false;
+    }
+    if (agentIsDisabledWithEnvVar()) {
+      LOGGER.warning("Agent was disabled with " + AGENT_ENABLED_ENV_VAR + " environment variable.");
+      return false;
+    }
+    if (agentIsAttached()) {
+      LOGGER.warning("Agent is already attached. It is not attached a second time.");
+      return false;
+    }
+    if (mainThreadCheckIsEnabled() && !isMainThread()) {
+      LOGGER.warning(
+          "Agent is not attached because runtime attachment was requested from main thread.");
+      return false;
+    }
+    return true;
+  }
+
+  private static boolean agentIsDisabledWithProp() {
+    String agentEnabledPropValue = System.getProperty(AGENT_ENABLED_PROPERTY);
+    return "false".equals(agentEnabledPropValue);
+  }
+
+  private static boolean agentIsDisabledWithEnvVar() {
+    String agentEnabledEnvVarValue = System.getenv(AGENT_ENABLED_ENV_VAR);
+    return "false".equals(agentEnabledEnvVarValue);
+  }
+
+  private static boolean agentIsAttached() {
+    try {
+      Class.forName("io.opentelemetry.javaagent.OpenTelemetryAgent", false, null);
+      return true;
+    } catch (ClassNotFoundException e) {
+      return false;
+    }
+  }
+
+  private static boolean mainThreadCheckIsEnabled() {
+    String mainThreadCheck = System.getProperty(MAIN_THREAD_CHECK_PROP);
+    return !"false".equals(mainThreadCheck);
```

Comment:
since this is an internal test-only flag, it may be worth inverting it (so the default is `false`, e.g. `...-check-disabled`), then can use `return Boolean.getBoolean(MAIN_THREAD_CHECK_PROP)` here

Reply:
For me the code would be a little less readable